api/src/context.rs

```diff
@@ -350,19 +350,30 @@ impl Context {
         address: AccountAddress,
         version: u64,
     ) -> Result<HashMap<StateKey, StateValue>> {
-        let mut iter = self.db.get_prefixed_state_value_iterator(
-            &StateKeyPrefix::from(address),
-            None,
-            version,
-        )?;
+        let mut iter = if !db_sharding_enabled(&self.node_config) {
```

Comment:
so many duplicated `if-else`s and `ensure!`s.. 

Why not `impl<T: DbReader> IndexerReader for T` and let the `Context` hold a `Arc<dyn IndexerReader>`?